apps/studio/data/prefetchers/project.$ref.editor.$id.tsx

```diff
@@ -0,0 +1,182 @@
+import { useProjectContext } from 'components/layouts/ProjectLayout/ProjectContext'
+import {
+  ForeignKeyConstraintsData,
+  prefetchForeignKeyConstraints,
+} from 'data/database/foreign-key-constraints-query'
+import { prefetchEncryptedColumns } from 'data/encrypted-columns/encrypted-columns-query'
+import { ENTITY_TYPE } from 'data/entity-types/entity-type-constants'
+import { Entity, prefetchEntityType } from 'data/entity-types/entity-type-query'
+import { ForeignTableData, prefetchForeignTable } from 'data/foreign-tables/foreign-table-query'
+import {
+  MaterializedViewData,
+  prefetchMaterializedView,
+} from 'data/materialized-views/materialized-view-query'
+import { prefetchTableRows } from 'data/table-rows/table-rows-query'
+import { prefetchTable, TableData } from 'data/tables/table-query'
+import { prefetchView, ViewData } from 'data/views/view-query'
+import { useFlag } from 'hooks/ui/useFlag'
+import { useRoleImpersonationStateSnapshot } from 'state/role-impersonation-state'
+import { TABLE_EDITOR_DEFAULT_ROWS_PER_PAGE } from 'state/table-editor'
+
+export function usePrefetchEditorTablePage() {
+  const router = useRouter()
+  const queryClient = useQueryClient()
+  const { project } = useProjectContext()
+  const roleImpersonationState = useRoleImpersonationStateSnapshot()
+  const tableEditorPrefetchingEnabled = useFlag('tableEditorPrefetching')
+
+  return useCallback(
+    ({ id: _id, filters, sorts }: { id?: string; filters?: Filter[]; sorts?: Sort[] }) => {
+      if (!tableEditorPrefetchingEnabled) return
+
+      const id = _id ? Number(_id) : undefined
+      if (!project || !id || isNaN(id)) return
+
+      // Prefetch the code
+      router.prefetch(`/project/${project.ref}/editor/${id}`)
+
+      // Prefetch the data
+      prefetchEntityType(queryClient, {
+        projectRef: project.ref,
+        connectionString: project.connectionString,
+        id,
+      })
+        .then((entity) => {
+          let promises: [
+            Promise<Entity | null>,
+            Promise<ForeignKeyConstraintsData>,
+            Promise<string[]>,
+            (
+              | Promise<TableData | ViewData | MaterializedViewData | ForeignTableData | undefined>
+              | undefined
+            ),
+          ] = [
+            Promise.resolve(entity),
+            prefetchForeignKeyConstraints(queryClient, {
+              projectRef: project.ref,
+              connectionString: project.connectionString,
+              schema: entity?.schema,
+            }),
+            prefetchEncryptedColumns(queryClient, {
+              projectRef: project.ref,
+              connectionString: project.connectionString,
+              schema: entity?.schema,
+              tableName: entity?.name,
+            }),
+            undefined,
+          ]
+
+          switch (entity?.type) {
+            case ENTITY_TYPE.TABLE:
+            case ENTITY_TYPE.PARTITIONED_TABLE:
+              promises[3] = prefetchTable(queryClient, {
```

Comment:
just a nit - i'd prefer immutability usually - should we just do promises.push() instead of reassigning index 3 instead? then we don't need the undefined as the last item in promises

Reply:
Yeah I'd rather use `.push()` too, the reason it's this way is to make typescript happy since I'm using a tuple instead of a regular array. So it knows `[3]` should be of type whatever `prefetchTable` returns